controller/persistence/eventual_eventer.go

```diff
@@ -0,0 +1,607 @@
+
+	// EventualEventProcessingListenerStartName is emitted for each function listener invoked
+	// on each event.
+	//
+	// Event arguments:
+	//	0 - an EventualEventProcessingListenerStart struct
+	EventualEventProcessingListenerStartName = events.EventName("EventualEventProcessingListenerStart")
+
+	// EventualEventProcessingListenerDoneName is emitted for each function listener after invocation
+	//
+	// Event arguments:
+	//	0 - an EventualEventProcessingListenerDone struct
+	EventualEventProcessingListenerDoneName = events.EventName("EventualEventProcessingListenerDone")
+
+	// EventualEventProcessingBatchDoneName is emitted after the last event processed in a batch.
+	//
+	// Event arguments:
+	//	0 - an EventualEventProcessingBatchDone struct
+	EventualEventProcessingBatchDoneName = events.EventName("EventualEventProcessingBatchDone")
+
+	// EventualEventProcessingDoneName is emitted as the last action during processing after
+	// all events and batches.
+	//
+	// Event arguments:
+	//	0 - an EventualEventProcessingDone struct
+	EventualEventProcessingDoneName = events.EventName("EventualEventProcessingDone")
+)
+
+// EventualEventerBbolt implements EventualEventer with a bbolt back storage mechanism.
+// Work is performed on a configurable basis via the Interval property in FIFO order.
+//
+// Events are stored in the following format:
+//		id   - CUID   - a monotonic reference id
+//      name - string - an event name, used for log output
+//      data - []byte - a string array of arguments
+type EventualEventerBbolt struct {
+	events.EventEmmiter
+	handlerMap        cmap.ConcurrentMap //eventName -> handler func(name, []byte)
+	Interval          time.Duration
+	closeNotify       <-chan struct{}
+	stopNotify        chan struct{}
+	trigger           chan struct{}
+	outstandingEvents *int64
+
+	waiters sync.Map //id -> chan struct{}
+
+	running    concurrenz.AtomicBoolean
+	batchSize  int
+	dbProvider DbProvider
+	store      EventualEventStore
+}
+
+var _ EventualEventer = &EventualEventerBbolt{}
+
+// NewEventualEventerBbolt creates a new bbolt backed asynchronous eventer that will check for new events at the given interval
+// or when triggered. On each interval/trigger, the number of events processed is determined by batchSize.
+func NewEventualEventerBbolt(dbProvider DbProvider, store EventualEventStore, interval time.Duration, batchSize int) *EventualEventerBbolt {
+	outstanding := int64(0)
+	return &EventualEventerBbolt{
+		EventEmmiter:      events.New(),
+		Interval:          interval,
+		dbProvider:        dbProvider,
+		store:             store,
+		batchSize:         batchSize,
+		trigger:           make(chan struct{}, 1),
+		handlerMap:        cmap.New(),
+		outstandingEvents: &outstanding,
+	}
+}
+
+func (a *EventualEventerBbolt) AddEventualEventWithCtx(ctx boltz.MutateContext, eventType string, data []byte) {
+	newId := cuid.New()
+	total := atomic.AddInt64(a.outstandingEvents, 1)
```

Comment:
I probably would have put the metric right in the store, but this should be ok too